code/game/machinery/firealarm.dm

```diff
@@ -306,104 +307,180 @@
-								set_machine_stat(machine_stat & ~BROKEN)
-							else
-								to_chat(user, span_notice("You pry out the circuit."))
-								new /obj/item/electronics/firealarm(user.loc)
-							buildstage = FIRE_ALARM_BUILD_NO_CIRCUIT
-							update_appearance()
-					return
-			if(0)
-				if(istype(tool, /obj/item/electronics/firealarm))
-					to_chat(user, span_notice("You insert the circuit."))
-					qdel(tool)
-					buildstage = FIRE_ALARM_BUILD_NO_WIRES
-					update_appearance()
-					return
-
-				else if(istype(tool, /obj/item/electroadaptive_pseudocircuit))
-					var/obj/item/electroadaptive_pseudocircuit/pseudoc = tool
-					if(!pseudoc.adapt_circuit(user, circuit_cost = 0.015 * STANDARD_CELL_CHARGE))
-						return
-					user.visible_message(span_notice("[user] fabricates a circuit and places it into [src]."), \
-					span_notice("You adapt a fire alarm circuit and slot it into the assembly."))
-					buildstage = FIRE_ALARM_BUILD_NO_WIRES
-					update_appearance()
-					return
-
-				else if(tool.tool_behaviour == TOOL_WRENCH)
-					user.visible_message(span_notice("[user] removes the fire alarm assembly from the wall."), \
-						span_notice("You remove the fire alarm assembly from the wall."))
-					var/obj/item/wallframe/firealarm/frame = new /obj/item/wallframe/firealarm()
-					frame.forceMove(user.drop_location())
-					tool.play_tool_sound(src)
-					qdel(src)
-					return
-	return ..()
+	loc.balloon_alert_to_viewers("removing circuit...")
+	if(!tool.use_tool(src, user, 2 SECONDS, volume = 50, extra_checks = CALLBACK(src, PROC_REF(state_callback), FIRE_ALARM_BUILD_NO_WIRES, TRUE)))
+		return ITEM_INTERACT_BLOCKING
+	if(machine_stat & BROKEN)
+		balloon_alert(user, "broken circuit removed")
+		set_machine_stat(machine_stat & ~BROKEN)
+	else
+		balloon_alert(user, "circuit removed")
+		new /obj/item/electronics/firealarm(user.drop_location())
+	buildstage = FIRE_ALARM_BUILD_NO_CIRCUIT
+	update_appearance()
+	return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/firealarm/crowbar_act_secondary(mob/living/user, obj/item/tool)
+	return crowbar_act(user, tool)
+
+/obj/machinery/firealarm/wrench_act(mob/living/user, obj/item/tool)
+	if(!panel_open)
+		return NONE
+	if(buildstage != FIRE_ALARM_BUILD_NO_CIRCUIT)
+		return NONE
+
+	loc.balloon_alert_to_viewers("[/obj/item/wallframe/firealarm::name] removed")
+	new /obj/item/wallframe/firealarm(user.drop_location())
+	tool.play_tool_sound(loc)
+	qdel(src)
+	return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/firealarm/wrench_act_secondary(mob/living/user, obj/item/tool)
+	return wrench_act(user, tool)
+
+/obj/machinery/firealarm/tool_act(mob/living/user, obj/item/tool, list/modifiers)
+	. = ..()
+	if(. & ITEM_INTERACT_ANY_BLOCKER)
+		return .
+	if(is_wire_tool(tool))
+		if(panel_open)
+			wires.interact(user)
+			return ITEM_INTERACT_SUCCESS
+
+		balloon_alert(user, "expose the wires first!")
+		return ITEM_INTERACT_BLOCKING
+	return NONE
```

Comment:
early returns instead of this?
```suggestion
	if(!is_wire_tool(tool))
		return NONE
	if(!panel_open)
		balloon_alert(user, "expose the wires first!")
		return ITEM_INTERACT_BLOCKING
	wires.interact(user)
	return ITEM_INTERACT_SUCCESS
```
beyond that, the `the` in the balloon alert may be unnecessary clutter to get the point across?
like might just make the balloon alert more unwieldy